sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -52,7 +54,15 @@ case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] {
       case (child, distribution) =>
         val numPartitions = distribution.requiredNumPartitions
           .getOrElse(conf.numShufflePartitions)
-        ShuffleExchangeExec(distribution.createPartitioning(numPartitions), child)
+
+        operator match {
+          // Ensure collectNullPartitionKeyMetrics is passed through via instance creation
+          case shj @ ShuffledHashJoinExec(_, _, LeftAnti, BuildRight, _, _, _, true)
+              if shj.right.fastEquals(child) =>
+            ShuffleExchangeExec(distribution.createPartitioning(numPartitions),
+              child, statSpecs = Seq(NullPartitionKeyStatisticsSpec))
+          case _ => ShuffleExchangeExec(distribution.createPartitioning(numPartitions), child)
+        }
```

Comment:
I feel this is not enough for passing `isNullAwareAntiJoin` information from `ShuffledHashJoinExec` to `ShuffleExchangeExec`. This only works when we need to insert a shuffle exchange at build side, but seems not cover when we don't need to insert a shuffle exchange (i.e. build side already partitioned on join keys).

Examples:

```
SELECT *
FROM t1
WHERE t1.key NOT IN (
  SELECT t2.key
  FROM t2
  DISTRIBUTE BY t2
)
```

or 

```
SELECT *
FROM t1
WHERE t1.key NOT IN (
  SELECT t2.key
  FROM t2 // t2 is bucketed table on column key
)
```

I feel we probably need a separate physical plan rule after `EnsureRequirements` to pass the null-aware anti join information from shuffled hash join to its child shuffle exchange operator.

Reply:
I will look into the example you were showing, never considered this case before, thanks.